internal/ethereum/exec_query.go

```diff
@@ -119,3 +130,63 @@ func (c *ethConnector) callTransaction(ctx context.Context, tx *ethsigner.Transa
 	return fftypes.JSONAnyPtrBytes(jsonData), "", nil
 
 }
+
+// processRevertReason returns under 3 different circumstances:
+// 1. non-empty string with nil error: valid reason has been successfully parsed
+// 2. non-empty string with non-nil error: error detail was present but failed to parse, string was raw data
+// 3. empty string with nil error: outputData is NOT an error detail data
+func processRevertReason(ctx context.Context, outputData ethtypes.HexBytes0xPrefix, errors []*abi.Entry) (string, error) {
+	// Check for a revert - we can determine it is calldata (with an error signature)
+	// that is returned by the fact the output is not a multiple of 32 (as all ABI encodings
+	// result in a multiple of 32 bytes) and has exactly 4 extra bytes for a function
+	// signature
+	if len(outputData)%32 == 4 {
+		signature := outputData[0:4]
+		if bytes.Equal(signature, defaultErrorID) {
+			errorInfo, err := defaultError.DecodeCallDataCtx(ctx, outputData)
+			if err == nil && len(errorInfo.Children) == 1 {
+				if strError, ok := errorInfo.Children[0].Value.(string); ok {
+					return strError, nil
+				}
+			}
+			log.L(ctx).Warnf("Invalid revert data: %s", outputData)
+		} else if len(errors) > 0 {
+			// check if the signature matches any of the declared custom error definitions
+			for _, e := range errors {
+				idBytes := e.FunctionSelectorBytes()
+				if bytes.Equal(signature, idBytes) {
+					err := formatCustomError(ctx, e, outputData)
+					if err == "" {
+						log.L(ctx).Warnf("Invalid revert data: %s", outputData)
+						break
+					}
+					return err, nil
+				}
+			}
+		}
+		return outputData.String(), fmt.Errorf("raw data returned")
```

Comment:
We shouldn't have any `fmt.Errorf()` - instead adding to the translated error file.

Reply:
this is a "transient" error as it simply signals to the caller that the string was going to be the raw data, then it gets thrown away. as such it doesn't need to be translated. but if the coding style requires translation, I'd be happy to add it. please let me know.

Reply:
A sentinel error would be a better pattern than a string constant, if you wouldn't mind.

Or simply a boolean `ok` return would be clearer in the case there's not an error for the caller to process, simply an indication that the data couldn't be extracted?

I would like us to have a linter that checks for `Errorf` anywhere in the code at some point.

Also reading the code caught me out because above `err != nil` didn't mean there was an error to process, but instead it meant that the function didn't complete its task.

Reply:
thanks for suggesting that. went with the boolean return value pattern.